File: senha (somente em C++)/senha2.cpp
// SENHA2.CPP

#include <iostream>
#include <cstring>
#include <cctype>
#if defined _WIN32 || defined _WIN64
  #include <windows.h>
#else
  #include <termios.h>
  #include <unistd.h>
  #include <cstdio>
#endif
using namespace std;

#if defined _WIN32 || defined _WIN64
  char getpwd(void)
  {
    char CARACTERE;
    HANDLE TECLADO;
    DWORD ESCRITA = 0, NUMCARAC;
    TECLADO = GetStdHandle(STD_INPUT_HANDLE);
    GetConsoleMode(TECLADO, &ESCRITA);
    SetConsoleMode(TECLADO, ESCRITA  & ~(ENABLE_LINE_INPUT | ENABLE_ECHO_INPUT));
    ReadConsole(TECLADO, &CARACTERE, 1, &NUMCARAC, NULL);
    SetConsoleMode(TECLADO, ESCRITA);
    return CARACTERE;
  }
#else
  int getpwd(void)
  {
    int CARACTERE;
    termios LEITURA;
    termios TECLADO;
    tcgetattr(STDIN_FILENO, &LEITURA);
    TECLADO = LEITURA;
    LEITURA.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &LEITURA);
    CARACTERE = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &TECLADO);
    return CARACTERE;
  }
#endif

char readpwd(char *SENHA)
{
  char CARACTERE;
  #if defined _WIN32 || defined _WIN64
    char CR = (char)  13;
    char BS = (char)   8;
  #else
    char CR = '\n';
    char BS = (char) 127;
  #endif
  int I = 0;
  do
    {
      CARACTERE = getpwd();
      if (isprint(CARACTERE))
        {
           cout << "*";
           SENHA[I] = CARACTERE;
           ++I;
        }
      else
        if (CARACTERE == BS && I != 0)
          {
             cout << "\b \b";
             SENHA[I] = '\0';
             --I;
          }
    }
  while(CARACTERE != CR);
  SENHA[I] = '\0';
  return *SENHA;
}

int main(void)
{

  char CADLOGIN[41], CADSENHA[11], ENTLOGIN[41], ENTSENHA[11];
  int TENTATIVAS = 3;

  cout << "CADASTRO DE ACESSO PARA USUARIO" << endl << endl;
  cout << "Entre nome para login ....: ";
  cin >> CADLOGIN;
  cin.ignore(80,'\n');

  cout << "Entre senha para login ...: ";
  readpwd(CADSENHA);

  cout << endl << endl;
  cout << "Usuario e senha cadastrados";
  cout << endl << endl;

  cout << "Tecle <Enter> para continuar... ";
  cin.get();
  cout << endl;

  do
    {

      cout << "Entre nome do usuario ....: ";
      cin >> ENTLOGIN;
      cin.ignore(80,'\n');

      cout << "Entre senha do usuario ...: ";
      readpwd(ENTSENHA);

      cout << endl << endl;
      if (not strcmp(CADLOGIN, ENTLOGIN) != 0 and not strcmp(CADSENHA, ENTSENHA) != 0)
        {
          cout << "ACESSO BEM SUCEDIDO" << endl;
          break;
        }
      else
        {
          --TENTATIVAS;
          if (TENTATIVAS != 0)
            {
              if (TENTATIVAS > 1)
                cout << "ERRO: SENHA/NOME - VOCE TEM " << TENTATIVAS << " TENTATIVAS.";
              else
                cout << "ERRO: SENHA/NOME - VOCE TEM " << TENTATIVAS << " TENTATIVA.";
              cout << endl << endl;
            }
        }
    }
  while (TENTATIVAS != 0);

  if (TENTATIVAS == 0)
    cout << "ACESSO NEGADO" << endl;

  cout << endl;
  cout << "Tecle <Enter> para encerrar... ";
  cin.get();
  return 0;
}

File: senha (somente em Pascal)/senha2.pas
program SENHA2;
uses
  {$IFDEF MSWINDOWS}
    Windows, SysUtils;
  {$ENDIF}
  {$IFDEF UNIX}
    {$linklib c}
    unixtype;
  {$ENDIF}

{$IFDEF MSWINDOWS}
  function getpwd : char;
  var
    CARACTERE : char;
    TECLADO : thandle;
    ESCRITA, NUMCARAC : dword;
  begin
    ESCRITA := 0;
    TECLADO := GetStdHandle(STD_INPUT_HANDLE);
    GetConsoleMode(TECLADO, @ESCRITA);
    SetConsoleMode(TECLADO, ESCRITA and not (ENABLE_LINE_INPUT or ENABLE_ECHO_INPUT));
    ReadConsole(TECLADO, @CARACTERE, 1, NUMCARAC, nil);
    SetConsoleMode(TECLADO, ESCRITA);
    getpwd := CARACTERE;
  end;
{$ENDIF}

{$IFDEF UNIX}
  function getpwd : integer;
  var
    CARACTERE : integer;
    LEITURA : termios;
    TECLADO : termios;
  begin
    tcgetattr(STDIN_FILENO, @LEITURA);
    TECLADO := LEITURA;
    LEITURA.c_lflag := LEITURA.c_lflag and not (ICANON or ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, @LEITURA);
    CARACTERE := getpwd;
    tcsetattr(STDIN_FILENO, TCSANOW, @TECLADO);
    getpwd := CARACTERE;
  end;
{$ENDIF}
  
function readpwd(var SENHA : string) : string;
var
  CARACTERE: char;
  CR, BS: char;
  I, TAMANHO: integer;
begin
  {$IFDEF MSWINDOWS}
    CR := #13;
    BS := #8;
  {$ENDIF}
  {$IFDEF UNIX}
    CR := #10;
    BS := #127;
  {$ENDIF}
  I := 0;
  TAMANHO := length(SENHA);
  repeat
    CARACTERE := getpwd();
    {$IFDEF MSWINDOWS}
      if (IsCharAlphaNumeric(CARACTERE)) then
    {$ENDIF}
    {$IFDEF UNIX}
      if (isalnum(ord(CARACTERE)) <> 0) then
    {$ENDIF}
      begin
        write('*');
        I := I + 1;
        if (I > TAMANHO) then
        begin
          SENHA := SENHA + CARACTERE;
          TAMANHO := TAMANHO + 1;
        end
        else
        begin
          SENHA[I] := CARACTERE;
        end;
      end
      else 
      if (CARACTERE = BS) and (I > 0) then
      begin
        write(#8' '#8);
        SENHA := copy(SENHA, 1, I - 1) + copy(SENHA, I + 1, TAMANHO);
        I := I - 1;
        TAMANHO := TAMANHO - 1;
      end;
  until (CARACTERE = CR);
  readpwd := SENHA;
end;
  
var
  CADLOGIN, CADSENHA, ENTLOGIN, ENTSENHA : string;
  TENTATIVAS : integer; 
  
begin
  TENTATIVAS := 3;
  
  writeln('CADASTRO DE ACESSO PARA USUARIO');
  writeln;
  write('Entre nome para login ....: ');
  readln(CADLOGIN);

  write('Entre senha para login ...: ');
  readpwd(CADSENHA);

  writeln;
  writeln;
  writeln('Usuario e senha cadastrados');
  writeln;

  write('Tecle <Enter> para continuar... ');
  readln;
  writeln;  

  repeat
    write('Entre nome do usuario ....: ');
    readln(ENTLOGIN);

    write('Entre senha do usuario ...: ');
    ENTSENHA := readpwd(ENTSENHA);
    writeln;
    writeln;

    if (CADLOGIN = ENTLOGIN) and (CADSENHA = ENTSENHA) then
      begin
        writeln('ACESSO BEM SUCEDIDO');
        break;
      end
    else
      begin
        TENTATIVAS := TENTATIVAS - 1;
        if (TENTATIVAS > 1) then
          writeln('ERRO: SENHA/NOME - VOCE TEM ', TENTATIVAS, ' TENTATIVAS.')
        else 
          writeln('ERRO: SENHA/NOME - VOCE TEM ', TENTATIVAS, ' TENTATIVA.');
        writeln;
      end;
  until not (TENTATIVAS <> 0);

  if (TENTATIVAS = 0) then
    writeln('ACESSO NEGADO');

  writeln;
  write('Tecle <Enter> para encerrar... ');
  readln;

end.
